### ISSUE TITLE
java_proto_library rule could not find imported .proto files from an external dependency

### ISSUE THREAD

**Issue by xfxyjwf** [bazelbuild/bazel#2039]
My project setup: two proto files in two repos, with one imports the other.

// importee
importee.proto

// importer
importer.proto

File content:
```
// importer/BUILD

proto_library(
    name = "importer_proto",
    srcs = ["importer.proto"],
    deps = ["@importee//:importee_proto"],
)

java_proto_library(
    name = "importer_proto_java",
    deps = [":importer_proto"],
)
```

```
// importer/WORKSPACE

local_repository(
    name = "importee",
    path = "path/to/importee",
)
```

When I run bazel build importer_proto_java, it gives out this error:

ERROR: /usr/local/google/home/xiaofeng/filer/workspace/git/bazel_proto_examples/import/BUILD:1:1: Generating Java proto_library //import:importer_proto failed: linux-sandbox failed: error executing command /usr/local/google/home/xiaofeng/.cache/bazel/_bazel_xiaofeng/25e42016a723de34add9af2f0eeffaf7/execroot/bazel_proto_examples/_bin/linux-sandbox ... (remaining 7 argument(s) skipped).
importee.proto: File not found.
import/importer.proto: Import "importee.proto" was not found or had errors.
import/importer.proto:8:12: "Importee" is not defined.

I created an example here:
https://github.com/xfxyjwf/bazel_proto_examples/tree/master/import

Just run "bazel build :all" in the import directory to reproduce.

bazel info release
```
release 0.3.2
```

**Comment by xfxyjwf** [comment]
@cgrushko

**Comment by cgrushko** [comment]
@laszlocsomor the issue is that when compiling a `.proto` file in the current repository, we need to read files in an external repository.

Is this possible by design?

**Comment by xfxyjwf** [comment]
I think it must be possible for the filegroup rule to work, right?

**Comment by cgrushko** [comment]
I can't find any method on Artifact that returns its absolute path.

Can filegroups cross repository boundaries?

**Comment by laszlocsomor** [comment]
> I can't find any method on Artifact that returns its absolute path.

That's by design. @kchodorow has been working hard to lock that down.

> Can filegroups cross repository boundaries?

Absolutely.

```
$ cat ./WORKSPACE
workspace(name = "foo")

local_repository(
    name = "bar",
    path = "/c/tempdir/repos/bar",
)


$ cat ./BUILD
filegroup(
    name = "files",
    srcs = ["@bar//:files"],
)

genrule(
    name = "foo-list-files",
    srcs = [":files"],
    outs = ["foo-list-files.txt"],
    cmd = "echo $(SRCS) > $@",
)


$ cat ../bar/WORKSPACE
workspace(name = "bar")


$ cat ../bar/BUILD
filegroup(
    name = "files",
    srcs = glob(["**/*"]),
    visibility = ["//visibility:public"],
)

genrule(
    name = "bar-list-files",
    srcs = [":files"],
    outs = ["bar-list-files.txt"],
    cmd = "echo $(SRCS) > $@",
)


$ bazel build //:foo-list-files @bar//:bar-list-files
...
WARNING: INFO: Build completed successfully, 1 total action


$ cat bazel-genfiles/foo-list-files.txt
external/bar/BUILD external/bar/WORKSPACE external/bar/hello.txt


$ cat bazel-genfiles/external/bar/bar-list-files.txt
external/bar/BUILD external/bar/WORKSPACE external/bar/hello.txt
```

**Comment by cgrushko** [comment]
@laszlocsomor @kchodorow we want to put `proto_library` rules in the `protobuf` repo, and depend on it from user-defined `proto_library` (that is, in the main repo). 

This requires that proto-compiler has access to `protobuf`'s `proto_library` when compiling main repo's `proto_library`.
 
Makes sense?

Why filegroups are allowed but other types of cross-over aren't?

**Comment by laszlocsomor** [comment]
They are.

I fiddled a bit with @xfxyjwf 's excellent repro (thanks for it! made debugging a lot easier), and found the culprit is the bad `import` line in `importer.proto`. It should be:
```
// import "importee.proto";  // original, broken
import "external/importee/importee.proto";  // working
```

Protobuf import statements must be fully qualified, relative to the current workspace (I believe), and relative to the importer workspace this is the path to the importee.

**Comment by cgrushko** [comment]
Huh. I never stopped to consider that "external/importee/importee.proto" is an acceptable path.

@xfxyjwf wdyt?

**Comment by laszlocsomor** [comment]
With my 20/20 hindsight it seems logical -- importing repositories is like `bind`-ing so it populates the `//external` package. But I only thought of this because of luckily stumbling upon similar stuff recently.

**Comment by xfxyjwf** [comment]
@laszlocsomor Using paths like "external/importee/importee.proto" is not an option because every proto_library must import the same file using exactly the same import path. If not, they won't compile. To demonstrate this, I added another import/importee/other.proto file which imports import/importee/importee.proto using 'import "importee.proto"', and then you can see the java_proto_library in import/BUILD file no longer compiles. This time the error message says:
```
external/importee/other.proto: Import "importee.proto" was not found or had errors.
```
Basically if in one repo a .proto file is imported using "foo/bar.proto", every other repo must imports the same .proto file using the same "foo/bar.proto" path. In a sense, the import path is part of the public API and can't be changed. For example, the descriptor.proto in protobuf library must be imported using 'import "google/protobuf/descriptor.proto"', regardless where the users put the descriptor.proto file.

**Comment by kchodorow** [comment]
#includes really shouldn't need the external/importee prefix. We have a test for this with the C++ rules (see https://github.com/bazelbuild/bazel/blob/master/src/test/shell/bazel/local_repository_test.sh#L335) but maybe the proto rules don't handle it correctly?

**Comment by cgrushko** [comment]
Any idea how this is handled? As I said above, there's no way to access an artifcat's path relative to its repository.

**Comment by cgrushko** [comment]
ping :)

**Comment by kchodorow** [comment]
Like so: https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibraryHelper.java#L1121-L1123
